Draw sample counts per line in source code

karupayun opened the pull request:
We want to include sample counts and percentages when showing source code
and disassembly view. This PR aims that. For that, we are creating a widget to 
be in the right, where we show that information. Also, we introduced a widget 
in the top with column titles.

This PR includes:

 - A right widget with information about sample counts 

 - A top-widget where we draw titles for every column ("Samples",
 "Function" & "Total").

 - Making the pop-up window wider

Related to http://b/159809330 from Source Code feature.

To test: Simply run CodeViewerDemo.

Screenshots with sample info, only to show how it looks:
Before:
<img width="500" alt="dialog before" src="https://user-images.githubusercontent.com/8610429/109789518-b69d1180-7c10-11eb-8a8f-50687b7a65b5.png">

After:
<img width="499" alt="source code" src="https://user-images.githubusercontent.com/8610429/109680968-4abd9900-7b7d-11eb-98c2-659cf5321cc1.png">

beckerhe reviewed (commented):
Very nice. Thanks Pablo. Looks good in general. I added some comments, mostly minor.

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -28,27 +29,37 @@ static const QColor kTextEditBackgroundColor{30, 30, 30};
 
       // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the line
       // number area.
+      // TODO: Is this really needed? Which update can break it?
```
You are going to address this TODO within this PR, right?

karupayun replied:
Yes

karupayun replied:
Actually, I want @beckerhe opinion here.

beckerhe replied:
I think this is needed. At least it doesn't hurt.

beckerhe replied:
I had a closer look into the qt docs. There are two situations here:

1. We need to handle scrolling - which is clear I guess
2. We need to handle content updates (changes of the source code)

The second doesn't really appear in our use-case because it's read-only. But It's still good style to handle it here. Just imagine we implement two different styles of line-break in the future. most editors allow to visually break a line when it is too long. We could have the same. And if the user toggles that every line with such a visual line break will have twice the height from before. And that's what's updated here.

beckerhe replied:
TLDR: Yes, we need that! It's our duty to handle that case, even if it is unlikely right now, that this code path is taken.

karupayun replied:
I changed. WDYT?

karupayun replied:
``  QRect titles_bar = rect;
      QRect line_number_rect = rect;
      QRect samples_info_rect = rect;

      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
      // bar's area
      titles_bar.setLeft(0);
      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
      top_bar_widget_.update(line_number_rect);

      line_number_rect.setLeft(0);
      line_number_rect.setWidth(left_sidebar_widget_.sizeHint().width());
      left_sidebar_widget_.update(line_number_rect);

      samples_info_rect.setLeft(0);
      samples_info_rect.setWidth(right_sidebar_widget_.sizeHint().width());
      right_sidebar_widget_.update(line_number_rect); ``

karupayun replied:
Reverted the change:
 
     ```
 QRect titles_bar = rect;
      QRect line_number_rect = rect;
      QRect samples_info_rect = rect;

      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
      // bar's area
      titles_bar.setLeft(0);
      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
      top_bar_widget_.update(line_number_rect);

      line_number_rect.setLeft(0);
      line_number_rect.setWidth(left_sidebar_widget_.sizeHint().width());
      left_sidebar_widget_.update(line_number_rect);

      samples_info_rect.setLeft(0);
      samples_info_rect.setWidth(right_sidebar_widget_.sizeHint().width());
      right_sidebar_widget_.update(line_number_rect);
```

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -72,15 +83,14 @@ Viewer::Viewer(QWidget* parent) : QPlainTextEdit(parent), left_sidebar_widget_{t
-  auto left_sidebar = contentsRect();
-  left_sidebar.setWidth(left_sidebar_widget_.sizeHint().width());
-  left_sidebar_widget_.setGeometry(left_sidebar);
+  UpdateSidebarPositions();
```
Why is it not also UpdateSidebarWidth?

karupayun replied:
Maybe I am wrong, but because they only change when the font-size change (so, after zooming in-out and not a resize event)?

florian-kuebler replied:
That might be the case. Basically wanted to point you to this, as before it has the "left_sidebar.SetWidth".

karupayun replied:
Yes, but we aren't changing the width there. If you read UpdateSidebarPositions we are also making that SetWidth.

This is code that was there before, but basically:
1 - Create a new left_sidebar with the entire windows size (  auto left_sidebar = contentsRect(); )
2- We set the width to the previous width (  left_sidebar.setWidth(left_sidebar_widget_.sizeHint().width()); )
3- We change our left_sidebar for the new one.

All these steps is to simply move the left_sidebar to the new position of the window without changing their size.

beckerhe replied:
`UpdateSidebarWidth` is only needed when something occured which leads to a different width of the sidebar. Since line numbers are shown there, one event that could cause a width-change is changing the shown code. A different number of lines might need more or less sidebar width due to more or fewer digits.

The same applies for changing the settings (enabling/disabling linenumbers/heatmap).

But in this event handler the only thing that actually changes is the size of the viewport, which means we might need to position the sidebar somewhere else, but we don't need to update its width.

Does that make sense?

florian-kuebler replied:
thanks

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -122,47 +132,143 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
 
-void Viewer::UpdateLeftSidebarWidth() {
+// For example, from a value = 0.5, we want to print "50.00 %"
+std::string PercentageFromDouble(double v) {
```
Nit: `PercentageStringFromDouble` or even `DoubleToPercentageString`?

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -122,47 +132,143 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
-void Viewer::UpdateLeftSidebarWidth() {
+// For example, from a value = 0.5, we want to print "50.00 %"
+std::string PercentageFromDouble(double v) {
+  std::clamp(v, 0., 1.);
```
Ideally we would have tests at least for the free functions. However, this can be done post-release, I guess.

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -194,4 +300,20 @@ void Viewer::HighlightCurrentLine() {
+int Viewer::WidthSampleCounterColumn() const {
+  const std::string kSampleColumnTitle = "Samples";
+  return StringWidthInPixels(fontMetrics(),
+                             kSampleColumnTitle);  // TODO: titles might have smaller font
```
Again, are you addressing this in this PR?

karupayun replied:
Yes

florian-kuebler reviewed (approved):
LGTM, some minor comments. Thanks!

vickyliu-go4it commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -28,27 +29,37 @@ static const QColor kTextEditBackgroundColor{30, 30, 30};
+                   &Viewer::DrawSampleCounters);
+
   const auto update_viewport_area = [&](const QRect& rect, int dy) {
     const auto update_caused_by_scroll = [](int dy) { return dy != 0; };
```
I'm not very clear about the purpose for using a lambda expression here. `update_cuased_by_scroll` seems only used in line 54. Why we are not using a boolean `update_cuased_by_scroll` here?

karupayun replied:
This was here before, and I also found it strange. I think that it was used to give the user a better explanation than a simply if(dy != 0). I changed it, What do you think?

vickyliu-go4it replied:
Thanks for the explanation and making the change :-) I understand the purpose for the readability. But just curious why not a simple `bool update_cuased_by_scroll = (dy != 0)` like we are using in other cases.

karupayun replied:
Nice suggestion!! I don't know why, but we are using it now :)

vickyliu-go4it reviewed (approved):
Nice PR! Thanks for that. I agree Florian's comments on the TODOs. If they will not be addressed in this PR, may be you can put the relevant bug in the TODOs.

LGTM!

karupayun commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -122,47 +132,143 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
+// For example, from a value = 0.5, we want to print "50.00 %"
+std::string PercentageFromDouble(double v) {
+  std::clamp(v, 0., 1.);
+  int printed_digits = static_cast<int>(round(v * 10000.));
+  return (std::to_string(printed_digits / 100) + "." + std::to_string(printed_digits % 100 / 10) +
+          std::to_string(printed_digits % 10) + " %");
```
Nice link. Will switch to QString!

karupayun commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -122,47 +132,143 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
+  left_sidebar.setWidth(left_sidebar_widget_.sizeHint().width());
+  left_sidebar_widget_.setGeometry(left_sidebar);
+
+  auto slider_width = QAbstractScrollArea().verticalScrollBar()->sizeHint().width();
```
Thanks! BTW, it is not solving the issue :(

beckerhe replied:
Too bad.

karupayun replied:
Solved by using MoveLeft instead of MoveRight! :)

beckerhe reviewed (approved):
Thank you for addressing my remarks.

karupayun commented:
I know that everyone did LGTM, but I'm asking for re-reviews. Sorry for that.

New changes: 
 - Addressed comments and TODO/bugs
 - We are using real data
 - We have a top_bar_widget now
 - The pop-up windows is bigger
 
 Change I need to do tomorrow:
  - Rebase on Henning's PR to include optional data (and not draw any number)
  
 Change I won't do:
  - Problem with zoom-in a lot (will set a bug tomorrow)

vickyliu-go4it reviewed (approved):
Very nice and clear PR! And I really like the way that you organize tasks and TODOs in the PR! Thanks for that!

florian-kuebler commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -198,4 +355,21 @@ void Viewer::HighlightCurrentLine() {
   setExtraSelections({selection});
 }
 
+int Viewer::WidthPercentageColumn() const {
```
This (and the next two functions) can be also static, right?

karupayun replied:
Yes, I need to add "static" to my mental checking list

beckerhe replied:
AFAIK, they need to be member function because the access `fontMetrics()`.

florian-kuebler replied:
Oh, sorry. Haven't noticed this one. Thanks!

karupayun replied:
Yes, Henning is right. Sorry for my comment and thanks Henning!

florian-kuebler reviewed (approved):
LGTM. Consider making the three functions static.

beckerhe commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -26,32 +27,57 @@ static const QColor kLineNumberBackgroundColor{50, 50, 50};
       line_number_rect.setWidth(left_sidebar_widget_.sizeHint().width());
       left_sidebar_widget_.update(line_number_rect);
+
+      samples_info_rect.setLeft(0);
```
I think the left coordinate needs to be the same as in `UpdateBarPositions`, so `total_width_without_scroll_bar - right_sidebar_width`

karupayun replied:
Of course, nice catch again!

beckerhe commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -26,32 +27,57 @@ static const QColor kLineNumberBackgroundColor{50, 50, 50};
+
+      // We keep the (vertical) y-coordinates and adjust the horizontal x-coordinates for the each
+      // bar's area
+      titles_bar.setLeft(0);
+      titles_bar.setWidth(top_bar_widget_.sizeHint().width());
+      top_bar_widget_.update(line_number_rect);
```
The title bar doesn't need to be updated at all, because a text change can't change anything here. I think you can remove it entirely.

beckerhe commented on src/CodeViewer/Viewer.cpp:
```diff
@@ -128,47 +192,156 @@ void Viewer::DrawLineNumbers(QPaintEvent* event) {
+    const int left = left_margin_.ToPixels(fontMetrics());
+
+    auto opt_num_samples_in_lines = code_report_->GetNumSamplesAtLine(block.blockNumber() + 1);
+    if (opt_num_samples_in_lines.has_value()) {
```
Nit: `if (!opt_num_samples_in_lines.has_value()) return;`

karupayun replied:
I wasn't sure there, but I will change it!

karupayun replied:
BTW, I should write _continue_ instead, not return

beckerhe commented on src/OrbitQt/orbitmainwindow.cpp:
```diff
@@ -1375,6 +1376,7 @@ void OrbitMainWindow::ShowSourceCode(const std::filesystem::path& file_path, siz
   orbit_code_viewer::Dialog code_viewer_dialog{this};
 
   code_viewer_dialog.SetEnableLineNumbers(true);
+  code_viewer_dialog.SetEnableSampleCounters(true);
```
I think this should only be enabled when `maybe_code_report.has_value()` is true. Can you move it down to line 1390?

karupayun replied:
nice catch!

beckerhe reviewed (approved):
Otherwise LGTM
